subprojects/base-services/src/main/java/org/gradle/internal/classloader/ClassLoaderUtils.java

```diff
@@ -16,17 +16,57 @@
 package org.gradle.internal.classloader;
 
 import org.gradle.api.Nullable;
+import org.gradle.internal.Cast;
 import org.gradle.internal.UncheckedException;
 import org.gradle.internal.concurrent.CompositeStoppable;
+import org.gradle.internal.reflect.JavaMethod;
+import org.gradle.internal.reflect.JavaReflectionUtil;
+import sun.misc.Unsafe;
 
 import java.io.IOException;
+import java.lang.reflect.Field;
 import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLConnection;
 
 public abstract class ClassLoaderUtils {
+
+    private static final Unsafe UNSAFE;
+
+    static {
+        try {
+            Field theUnsafe = Unsafe.class.getDeclaredField("theUnsafe");
+            theUnsafe.setAccessible(true);
+            UNSAFE = (Unsafe) theUnsafe.get(null);
+        } catch (NoSuchFieldException e) {
+            throw new RuntimeException(e);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static final JavaMethod<ClassLoader, Package[]> GET_PACKAGES_METHOD;
+    private static final JavaMethod<ClassLoader, Package> GET_PACKAGE_METHOD;
+
+    static {
+        GET_PACKAGES_METHOD = getMethodWithFallback(Package[].class, new Class[0], "getDefinedPackages", "getPackages");
+        GET_PACKAGE_METHOD = getMethodWithFallback(Package.class, new Class[] {String.class}, "getDefinedPackage", "getPackage");
+    }
+
+    private static <T> JavaMethod<ClassLoader, T> getMethodWithFallback(Class<T> clazz, Class<?>[] params, String firstChoice, String fallback) {
+        JavaMethod<ClassLoader, T> method;
+        try {
+            method = JavaReflectionUtil.method(ClassLoader.class, clazz, firstChoice, params);
+        } catch (Throwable e) {
```

Comment:
In that case you can do 

```
catch (VirtualMachineError e) {
 throw e
} //other catch blocks here
```

That way you won't swallow OOME's and similar unrecoverable problems.